Fix #4760 In Terms of Service the website should be displayed as link and should be clickable.

Tejas-67 opened the pull request:
## Explanation
Fix #4760 In Terms of Service the website should be displayed as link and should be clickable
This Solution uses SpannableString to solve the the issue. We get the url from html spannable using regex and then set a span around it to make the link clickable.
## Essential Checklist
<!-- Please tick the relevant boxes by putting an "x" in them. -->
- [x] The PR title and explanation each start with "Fix #bugnum: " (If this PR fixes part of an issue, prefix the title with "Fix part of #bugnum: ...".)
- [x] Any changes to [scripts/assets](https://github.com/oppia/oppia-android/tree/develop/scripts/assets) files have their rationale included in the PR explanation.
- [x] The PR follows the [style guide](https://github.com/oppia/oppia-android/wiki/Coding-style-guide).
- [x] The PR does not contain any unnecessary code changes from Android Studio ([reference](https://github.com/oppia/oppia-android/wiki/Guidance-on-submitting-a-PR#undo-unnecessary-changes)).
- [x] The PR is made from a branch that's **not** called "develop" and is up-to-date with "develop".
- [x] The PR is **assigned** to the appropriate reviewers ([reference](https://github.com/oppia/oppia-android/wiki/Guidance-on-submitting-a-PR#clarification-regarding-assignees-and-reviewers-section)).

## For UI-specific PRs only
<!-- Delete these section if this PR does not include UI-related changes. -->
If your PR includes UI-related changes, then:
- Add screenshots for portrait/landscape for both a tablet & phone of the before & after UI changes
- For the screenshots above, include both English and pseudo-localized (RTL) screenshots (see [RTL guide](https://github.com/oppia/oppia-android/wiki/RTL-Guidelines))
- Add a video showing the full UX flow with a screen reader enabled (see [accessibility guide](https://github.com/oppia/oppia-android/wiki/Accessibility-A11y-Guide))
- For PRs introducing new UI elements or color changes, both light and dark mode screenshots must be included
- Add a screenshot demonstrating that you ran affected Espresso tests locally & that they're passing
## Demo Video
[oppia-.webm](https://github.com/oppia/oppia-android/assets/110051718/2d5b5bf5-3aad-43da-8b41-7146620e75ab)

Wintersoldi3r-2004 committed:
Fix oppia#4760 In Terms of Service the website should be displayed as link and should be clickable.

adhiamboperes commented on app/src/main/java/org/oppia/android/app/policies/PoliciesFragmentPresenter.kt:
```diff
@@ -49,35 +60,64 @@ class PoliciesFragmentPresenter @Inject constructor(
-    )
+      val spannableString = SpannableString(HtmlCompat.fromHtml(policyDescription, HtmlCompat.FROM_HTML_MODE_LEGACY))
+      val uri = "https://www.oppia.org"
+      val spanStart = policyDescription.indexOf(uri) - 4
```
Why are we using a fixed value of 4 here? What does it represent?

Tejas-67 replied:
the spanStart was returning the index of the fifth character in the link i.e 's' from https://www.oppia.org. Only fix was to subtract 4 from the starting index as i couldn't figure out why.

adhiamboperes replied:
This seems brittle, what happens if the uri were to change? Can you look for an inbuildt function for extracting the first part of a uri? You could also use pattern matching/regex.

Tejas-67 replied:
Hey @adhiamboperes  I have fixed and upated the PR. Please check.

adhiamboperes commented on app/src/main/java/org/oppia/android/app/policies/PoliciesFragmentPresenter.kt:
```diff
@@ -49,35 +60,64 @@ class PoliciesFragmentPresenter @Inject constructor(
+      }
+      spannableString.setSpan(clickableSpan, spanStart, spanEnd, Spanned.SPAN_EXCLUSIVE_EXCLUSIVE)
+      binding.policyDescriptionTextView.text = spannableString
+      binding.policyDescriptionTextView.movementMethod = LinkMovementMethod.getInstance()
+      binding.policyWebLinkTextView.text = htmlParserFactory.create(
+        gcsResourceName = "",
+        entityType = "",
+        entityId = "",
+        imageCenterAlign = false,
+        customOppiaTagActionListener = null,
+        resourceHandler.getDisplayLocale()
+      ).parseOppiaHtml(
+        policyWebLink,
+        binding.policyWebLinkTextView,
+        supportsLinks = true,
+        supportsConceptCards = false
+      )
+    }
```
DRY: This code is identical to the if branch above. Can you reduce duplication by moving it out of the branches, or to a function?

Tejas-67 replied:
Sure

adhiamboperes reviewed (changes requested):
Thanks @Tejas-67!

Please include a video demo of your solution working as expected.

Once you have addressed the comments, please assig this back to me. See [wiki section](https://github.com/oppia/oppia-android/wiki/Guidance-on-submitting-a-PR#clarification-regarding-assignees-and-reviewers-section).

Wintersoldi3r-2004 committed:
Refactor link detection in policyDescription

- Replace brittle string manipulation with robust regex matching using .toRegex().find()
- Introduce new helper functions for code simplification

adhiamboperes commented on app/src/main/java/org/oppia/android/app/policies/PoliciesFragmentPresenter.kt:
```diff
@@ -79,13 +112,15 @@ class PoliciesFragmentPresenter @Inject constructor(
-      PolicyType.TERMS_OF_SERVICE ->
-        (activity as RouteToPoliciesListener).onRouteToPolicies(PolicyPage.TERMS_OF_SERVICE)
-    }
+  private fun updatePolicyDescriptionTvWithParsedHtml(binding: PoliciesFragmentBinding, policyDescription: String){
```
Nit: Naming.
```suggestion
  private fun updatePolicyTextViewWithParsedHtml(binding: PoliciesFragmentBinding, policyDescription: String){
```

Tejas-67 replied:
@adhiamboperes  i have made the specified changes, please check.

adhiamboperes reviewed (approved):
Thanks @Tejas-67, I have added a nit, but nothing blocking. Please fix and reassign.

Wintersoldi3r-2004 committed:
Refractor functions to set text in textviews

adhiamboperes commented:
Thanks @Tejas-67. Could you please fix the failing tests? See https://github.com/oppia/oppia-android/wiki/Interpreting-CI-Results for finding out which tests are failing.

Tejas-67 commented:
Hey @adhiamboperes , fixed some lint issues that was causing the tests to fail. Other tests failing because the link to navigate to privacy policy page from terms of service page is not working. A possible fix would be to add another span for the link. Should i go ahead with this fix or raise a issue after this PR is merged?

adhiamboperes commented:
> Hey @adhiamboperes , fixed some lint issues that was causing the tests to fail. Other tests failing because the link to navigate to privacy policy page from terms of service page is not working. A possible fix would be to add another span for the link. Should i go ahead with this fix or raise a issue after this PR is merged?

@Tejas-67, all existing tests must pass for the PR to be merged. Can you identify why the links are not working? What has changed?

Tejas-67 commented:
@adhiamboperes , The links are not working because the text is enclosed in <oppia-noninteractive-policy link=\"privacy\">Privacy Policy</oppia-noninteractive-policy> and html.fromHtml doesn't recognize this. One fix could be to to wrap the text with another span.

adhiamboperes reviewed (changes requested):
> @adhiamboperes , The links are not working because the text is enclosed in <oppia-noninteractive-policy link="privacy">Privacy Policy and html.fromHtml doesn't recognize this. One fix could be to to wrap the text with another span.


I think this was working before your fix, right? It is worth taking a step back and finding a different solution. Take a look at [this class](https://github.com/oppia/oppia-android/blob/develop/utility/src/main/java/org/oppia/android/util/parser/html/CustomHtmlContentHandler.kt) that handles custom html tags. I believe it has a fromHtml() function that is used in the  [HtmlParser.kt class](https://github.com/oppia/oppia-android/blob/develop/utility/src/main/java/org/oppia/android/util/parser/html/HtmlParser.kt), which in [computeCustomTagHandlers()](https://github.com/oppia/oppia-android/blob/913cf5fba3ccce8e8befcd3684d612aed66a36a4/utility/src/main/java/org/oppia/android/util/parser/html/HtmlParser.kt#L134), parses the policy tag.

adhiamboperes commented:
@Tejas-67, are you still working on this PR?

Tejas-67 commented:
@adhiamboperes  Yes, I am working on it.

Wintersoldi3r-2004 committed:
FIX: PrivacyPolicy navigation issue.

Tejas-67 commented:
Hey @adhiamboperes , please check the new code.

Wintersoldi3r-2004 committed:
Remove unused imports.

Tejas-67 commented:
Hey @adhiamboperes, do I need to raise a new PR for this issue?

adhiamboperes commented on app/src/main/res/values/terms_of_service.xml:
```diff
@@ -236,4 +236,4 @@
     </ul>
 </]]></string>
-</resources>
+</resources>
```
Please insert new line at the end of the file

adhiamboperes commented on utility/src/main/java/org/oppia/android/util/parser/html/HtmlParser.kt:
```diff
@@ -86,7 +88,7 @@ class HtmlParser private constructor(
       val modifiedHtmlContent = rawString.replace(regex) {
         val oppiaImageTag = it.value
-        """<div style="text-align: center;">$oppiaImageTag</div>"""
+        """<div style="centtext-align: er;">$oppiaImageTag</div>"""
```
I'm sure this was an erronuous change, please undo.

adhiamboperes commented on utility/src/main/java/org/oppia/android/util/parser/html/HtmlParser.kt:
```diff
@@ -117,7 +119,8 @@ class HtmlParser private constructor(
     if (supportsLinks) {
       htmlContentTextView.movementMethod = LinkMovementMethod.getInstance()
-      LinkifyCompat.addLinks(htmlContentTextView, Linkify.WEB_URLS)
+      LinkifyCompat.addLinks(htmlContentTextView, Linkify.ALL)
+      htmlContentTextView.linksClickable = true
```
I tested the code manually and found that it is safe to remove this change.

adhiamboperes commented on utility/src/main/java/org/oppia/android/util/parser/html/HtmlParser.kt:
```diff
@@ -128,6 +131,15 @@ class HtmlParser private constructor(
       htmlContent, imageGetter, computeCustomTagHandlers(supportsConceptCards, htmlContentTextView)
     )
 
+    val websiteLink = "https://www.oppia.org"
+    val start = htmlSpannable.indexOf(websiteLink)
+    val end = start + websiteLink.length
+    if (start != -1) {
+      val urlSpan = URLSpan(websiteLink)
+      htmlSpannable.setSpan(
+        urlSpan, start, end, Spanned.SPAN_EXCLUSIVE_INCLUSIVE
+      )
+    }
```
This is the only code that actually needs to be added to this PR to fix the issue. I have a suggestion on how to improve it so that the fix applies to all links, not just the Oppia website. 
```suggestion
    val urlPattern = Patterns.WEB_URL
    val matcher = urlPattern.matcher(htmlSpannable)

    while (matcher.find()) {
      val start = matcher.start()
      val end = matcher.end()
      val url = htmlSpannable.subSequence(start, end).toString()

      val urlSpan = URLSpan(url)
      htmlSpannable.setSpan(urlSpan, start, end, Spanned.SPAN_EXCLUSIVE_INCLUSIVE)
    }
```

adhiamboperes reviewed (approved):
Thanks @Tejas-67! (Sorry for the delay in review here).

Based on previous conversation, you do not need to close a PR and open a new one, when undoing unwanted changes, you can simply create new commits after you have changed code back to the way it was, and  push to the branch.

(However when working on a different issue, you need to create a new branch from develop).

Our wiki has a good set of instructions that can help. https://github.com/oppia/oppia-android/wiki/Guidance-on-submitting-a-PR. Please let me know if you have further questions.

adhiamboperes reviewed (changes requested):
Sorry, meant to request changes as per my comments.

Wintersoldi3r-2004 committed:
Applied requested changes from PR review

Tejas-67 commented:
@adhiamboperes please check.

adhiamboperes reviewed (approved):
Thanks @Tejas-67!

adhiamboperes committed:
Merge branch 'develop' into develop-tejas
